lib/client/mfa.go

```diff
@@ -18,72 +18,74 @@ package client
 
 import (
 	"context"
-	"errors"
 	"fmt"
-	"io"
 	"os"
 	"strings"
 	"sync"
+	"sync/atomic"
 
 	"github.com/gravitational/teleport/api/client/proto"
-	"github.com/gravitational/teleport/lib/utils/prompt"
 	"github.com/gravitational/trace"
 
 	wanlib "github.com/gravitational/teleport/lib/auth/webauthn"
 	wancli "github.com/gravitational/teleport/lib/auth/webauthncli"
 )
 
-type (
-	OTPPrompt func(ctx context.Context, out io.Writer, in prompt.Reader, question string) (string, error)
-	WebPrompt func(
-		ctx context.Context,
-		origin, user string,
-		assertion *wanlib.CredentialAssertion,
-		prompt wancli.LoginPrompt) (*proto.MFAAuthenticateResponse, string, error)
-)
+// promptWebauthn provides indirection for tests.
+var promptWebauthn = wancli.Login
 
-// PlatformPrompt groups functions that prompt the user for inputs.
-// It's purpose is to allow tests to replace actual user prompts with other
-// functions.
-type PlatformPrompt struct {
-	// OTP is the OTP prompt function.
-	OTP OTPPrompt
-	// Webauthn is the WebAuth prompt function.
-	Webauthn WebPrompt
+type stdinRead struct {
+	value string
+	err   error
 }
 
-func (pp *PlatformPrompt) Reset() *PlatformPrompt {
-	pp.Swap(prompt.Input, wancli.Login)
-	return pp
+// stdinHijack hijacks stdin for a single password-like read.
+// After startRead is called the read will be sent to the C channel.
+type stdinHijack struct {
+	C       chan stdinRead
+	started int32
 }
 
-func (pp *PlatformPrompt) Swap(otp OTPPrompt, web WebPrompt) {
-	pp.OTP = otp
-	pp.Webauthn = web
+func (h *stdinHijack) startRead() {
+	if !atomic.CompareAndSwapInt32(&h.started, 0, 1) {
+		return // Already started
+	}
+	h.C = make(chan stdinRead)
+	go func() {
+		value, err := PasswordFromConsole()
+		h.C <- stdinRead{value: value, err: err}
+	}()
 }
 
-var prompts = (&PlatformPrompt{}).Reset()
-
-type noopPrompt struct{}
+// mfaPrompt implements wancli.LoginPrompt for MFA logins.
+// In most cases authenticators shouldn't require PINs or additional touches for
+// MFA, but the implementation exists in case we find some unusual
+// authenticators out there.
+type mfaPrompt struct {
+	// stdin and cancel are used in case of PIN reads.
+	stdin  *stdinHijack
+	cancel context.CancelFunc
```

Comment:
maybe rename it to `cancelOTP` to make it clearer?

Reply:
Done, although I used `otpCancel`.